File: src/stores/projects.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import type { Project } from "@/types";
import { invoke } from "@tauri-apps/api/core";

export const useProjectsStore = defineStore("projects", () => {
  const projects = ref<Project[]>([]);
  const loading = ref(false);
  const error = ref<string | null>(null);

  const recentProjects = computed(() => {
    return [...projects.value]
      .filter((p) => p.lastOpened)
      .sort(
        (a, b) =>
          new Date(b.lastOpened!).getTime() - new Date(a.lastOpened!).getTime()
      )
      .slice(0, 5);
  });

  const projectById = computed(() => {
    return (id: string) => projects.value.find((p) => p.id === id);
  });

  async function loadProjects() {
    console.log("[ProjectsStore] Loading projects...");
    loading.value = true;
    error.value = null;
    try {
      projects.value = await invoke<Project[]>("get_projects");
      console.log("[ProjectsStore] Loaded", projects.value.length, "projects");
    } catch (e) {
      error.value = String(e);
      console.error("[ProjectsStore] Failed to load projects:", e);
    } finally {
      loading.value = false;
    }
  }

  async function addProject(project: Omit<Project, "id" | "createdAt">) {
    console.log("[ProjectsStore] Adding project:", project.name);
    try {
      const newProject = await invoke<Project>("add_project", { project });
      projects.value.push(newProject);
      console.log("[ProjectsStore] Project added with ID:", newProject.id);
      return newProject;
    } catch (e) {
      error.value = String(e);
      console.error("[ProjectsStore] Failed to add project:", e);
      throw e;
    }
  }

  async function updateProject(id: string, updates: Partial<Project>) {
    console.log("[ProjectsStore] Updating project:", id, "with:", Object.keys(updates));
    try {
      const updated = await invoke<Project>("update_project", { id, updates });
      const index = projects.value.findIndex((p) => p.id === id);
      if (index !== -1) {
        projects.value[index] = updated;
      }
      console.log("[ProjectsStore] Project updated successfully");
      return updated;
    } catch (e) {
      error.value = String(e);
      console.error("[ProjectsStore] Failed to update project:", e);
      throw e;
    }
  }

  async function deleteProject(id: string) {
    console.log("[ProjectsStore] Deleting project:", id);
    try {
      await invoke("delete_project", { id });
      projects.value = projects.value.filter((p) => p.id !== id);
      console.log("[ProjectsStore] Project deleted successfully");
    } catch (e) {
      error.value = String(e);
      console.error("[ProjectsStore] Failed to delete project:", e);
      throw e;
    }
  }

  async function openProject(id: string) {
    console.log("[ProjectsStore] Opening project:", id);
    try {
      await invoke("open_project", { id });
      const project = projects.value.find((p) => p.id === id);
      if (project) {
        project.lastOpened = new Date().toISOString();
      }
      console.log("[ProjectsStore] Project opened successfully");
    } catch (e) {
      error.value = String(e);
      console.error("[ProjectsStore] Failed to open project:", e);
      throw e;
    }
  }

  return {
    projects,
    loading,
    error,
    recentProjects,
    projectById,
    loadProjects,
    addProject,
    updateProject,
    deleteProject,
    openProject,
  };
});

File: src-tauri/src/commands/projects.rs
use crate::config::{self, NewProject, Project, Environment, Workspace, Command};
use crate::error::{CmdrError, Result};
use std::collections::HashMap;

#[tauri::command]
pub fn get_projects() -> Result<Vec<Project>> {
    log::info!("[Projects] Getting all projects");
    let projects = config::load_projects()?;
    log::info!("[Projects] Returning {} projects", projects.len());
    Ok(projects)
}

#[tauri::command]
pub fn add_project(project: NewProject) -> Result<Project> {
    log::info!("[Projects] Adding new project: {}", project.name);
    let project = project.into_project();
    config::save_project(&project)?;
    log::info!("[Projects] Project added with ID: {}", project.id);
    Ok(project)
}

#[tauri::command]
pub fn update_project(id: String, updates: HashMap<String, serde_json::Value>) -> Result<Project> {
    log::info!("[Projects] Updating project: {}", id);
    log::debug!("[Projects] Update fields: {:?}", updates.keys().collect::<Vec<_>>());

    let projects = config::load_projects()?;
    let mut project = projects
        .into_iter()
        .find(|p| p.id == id)
        .ok_or_else(|| {
            log::error!("[Projects] Project not found: {}", id);
            CmdrError::ProjectNotFound(id.clone())
        })?;

    // Apply updates
    if let Some(name) = updates.get("name").and_then(|v| v.as_str()) {
        log::debug!("[Projects] Updating name: {} -> {}", project.name, name);
        project.name = name.to_string();
    }
    if let Some(path) = updates.get("path").and_then(|v| v.as_str()) {
        log::debug!("[Projects] Updating path: {} -> {}", project.path, path);
        project.path = path.to_string();
    }
    if let Some(description) = updates.get("description") {
        log::debug!("[Projects] Updating description");
        project.description = description.as_str().map(|s| s.to_string());
    }
    if let Some(tags) = updates.get("tags").and_then(|v| v.as_array()) {
        log::debug!("[Projects] Updating tags");
        project.tags = tags
            .iter()
            .filter_map(|t| t.as_str().map(|s| s.to_string()))
            .collect();
    }
    if let Some(last_opened) = updates.get("lastOpened").and_then(|v| v.as_str()) {
        log::debug!("[Projects] Updating lastOpened");
        project.last_opened = Some(last_opened.to_string());
    }
    if let Some(environments) = updates.get("environments") {
        log::debug!("[Projects] Updating environments");
        match serde_json::from_value::<HashMap<String, Environment>>(environments.clone()) {
            Ok(envs) => project.environments = envs,
            Err(e) => log::error!("[Projects] Failed to parse environments: {} - value: {:?}", e, environments),
        }
    }
    if let Some(workspaces) = updates.get("workspaces") {
        log::debug!("[Projects] Updating workspaces");
        match serde_json::from_value::<Vec<Workspace>>(workspaces.clone()) {
            Ok(ws) => project.workspaces = ws,
            Err(e) => log::error!("[Projects] Failed to parse workspaces: {} - value: {:?}", e, workspaces),
        }
    }
    if let Some(commands) = updates.get("commands") {
        log::debug!("[Projects] Updating commands");
        match serde_json::from_value::<Vec<Command>>(commands.clone()) {
            Ok(cmds) => project.commands = cmds,
            Err(e) => log::error!("[Projects] Failed to parse commands: {} - value: {:?}", e, commands),
        }
    }

    config::save_project(&project)?;
    log::info!("[Projects] Project updated successfully: {}", project.name);
    Ok(project)
}

#[tauri::command]
pub fn delete_project(id: String) -> Result<()> {
    log::info!("[Projects] Deleting project: {}", id);
    config::delete_project_file(&id)?;
    log::info!("[Projects] Project deleted successfully");
    Ok(())
}

#[tauri::command]
pub fn open_project(id: String) -> Result<()> {
    log::info!("[Projects] Opening project: {}", id);

    let projects = config::load_projects()?;
    let mut project = projects
        .into_iter()
        .find(|p| p.id == id)
        .ok_or_else(|| {
            log::error!("[Projects] Project not found: {}", id);
            CmdrError::ProjectNotFound(id.clone())
        })?;

    log::info!("[Projects] Found project: {} at path: {}", project.name, project.path);

    // Update last opened timestamp
    project.last_opened = Some(chrono::Utc::now().to_rfc3339());
    config::save_project(&project)?;

    // Open in default file manager/IDE
    // For now, just open the folder
    #[cfg(target_os = "macos")]
    {
        log::info!("[Projects] Opening folder in Finder: {}", project.path);
        match std::process::Command::new("open")
            .arg(&project.path)
            .spawn() {
            Ok(_) => log::info!("[Projects] Folder opened successfully"),
            Err(e) => log::error!("[Projects] Failed to open folder: {}", e),
        }
    }

    #[cfg(target_os = "windows")]
    {
        log::info!("[Projects] Opening folder in Explorer: {}", project.path);
        match std::process::Command::new("explorer")
            .arg(&project.path)
            .spawn() {
            Ok(_) => log::info!("[Projects] Folder opened successfully"),
            Err(e) => log::error!("[Projects] Failed to open folder: {}", e),
        }
    }

    #[cfg(target_os = "linux")]
    {
        log::info!("[Projects] Opening folder with xdg-open: {}", project.path);
        match std::process::Command::new("xdg-open")
            .arg(&project.path)
            .spawn() {
            Ok(_) => log::info!("[Projects] Folder opened successfully"),
            Err(e) => log::error!("[Projects] Failed to open folder: {}", e),
        }
    }

    Ok(())
}
